### PR TITLE
Adding all_reduce facility to collectives module

### PR DESCRIPTION
**Pull request by hkaiser**
- flyby: minor fixes to all_to_all

### CODE REVIEW

File: libs/collectives/include/hpx/collectives/all_reduce.hpp [hunk 1]
```diff
@@ -0,0 +1,512 @@
...
+
+                // this is a one-shot object (generations counters are not
+                // supported), unregister ourselves (but only once)
+                if (!name.empty())
+                {
+                    hpx::unregister_with_basename(name, site_);
+                }
+
+                return hpx::parallel::reduce(hpx::parallel::execution::par,
+                    data.begin(), data.end(), std::forward<F>(op));
```
Comment (msimberg) [inline libs/collectives/include/hpx/collectives/all_reduce.hpp]: Where is `op` defined? https://circleci.com/gh/STEllAR-GROUP/hpx/102137?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-checks-link

Comment (hkaiser) [reply libs/collectives/include/hpx/collectives/all_reduce.hpp]: This is an oversight, the function is not needed anymore. Will fix.

**Comment by hkaiser** [comment]
@msimberg this is good to go now. Would you have the time to review?

File: libs/collectives/include/hpx/collectives/all_reduce.hpp [hunk 2]
```diff
@@ -0,0 +1,486 @@
...
+        template <typename T>
+        class all_reduce_server
+          : public hpx::components::component_base<all_reduce_server<T> >
+        {
+            using mutex_type = lcos::local::spinlock;
+
+        public:
+            all_reduce_server() //-V730
+            {
+                HPX_ASSERT(false);  // shouldn't ever be called
```
[the remark this answers is not in the document]
Comment (hkaiser) [reply libs/collectives/include/hpx/collectives/all_reduce.hpp]: Components need to be default-constructable, so deleting the default constructor is not an option, unfortunately.

**Pull request merged**